src/htmx.js

```diff
@@ -3525,43 +3534,56 @@ return (function () {
-                if (event.state && event.state.htmx) {
-                    restoreHistory();
-                    forEach(restoredElts, function(elt){
-                        triggerEvent(elt, 'htmx:restored', {
-                            'document': getDocument(),
-                            'triggerEvent': triggerEvent
+        var started = false;
+        function start() {
+            if (started) {
+                logWarn('htmx has already been initialized on this page. Calling htmx.start() more than once may cause problems.')
+            }
+            started = true;
+
+            ready(function() {
+                mergeMetaConfig();
+                insertIndicatorStyles();
+                var body = getDocument().body;
+                processNode(body);
+                var restoredElts = getDocument().querySelectorAll(
+                    "[hx-trigger='restored'],[data-hx-trigger='restored']"
+                );
+                body.addEventListener("htmx:abort", function (evt) {
+                    var target = evt.target;
+                    var internalData = getInternalData(target);
+                    if (internalData && internalData.xhr) {
+                        internalData.xhr.abort();
+                    }
+                });
+                var originalPopstate = window.onpopstate;
+                window.onpopstate = function (event) {
+                    if (event.state && event.state.htmx) {
+                        restoreHistory();
+                        forEach(restoredElts, function(elt){
+                            triggerEvent(elt, 'htmx:restored', {
+                                'document': getDocument(),
+                                'triggerEvent': triggerEvent
+                            });
                         });
-                    });
-                } else {
-                    if (originalPopstate) {
-                        originalPopstate(event);
+                    } else {
+                        if (originalPopstate) {
+                            originalPopstate(event);
+                        }
                     }
-                }
-            };
-            setTimeout(function () {
-                triggerEvent(body, 'htmx:load', {}); // give ready handlers a chance to load up before firing this event
-                body = null; // kill reference for gc
-            }, 0);
-        })
+                };
+                setTimeout(function () {
+                    triggerEvent(body, 'htmx:load', {}); // give ready handlers a chance to load up before firing this event
+                    body = null; // kill reference for gc
+                }, 0);
+            });
+        }
+
+        // do not initialize htmx automatically if it is executed in module scope
+        // or the script element has the attribute async or defer
+        if (document.currentScript && !document.currentScript.async && !document.currentScript.defer) {
```

Comment:
Why not make this a param in the [config](https://htmx.org/docs/#config) that defaults to true, instead of that defer/async attributes check ?
You would keep the ability to disable htmx auto start when it becomes an issue, but I'd argue that since it's an issue in **some** cases, the auto initialization should be the default
So the behaviour wouldn't change for existing apps using htmx as defer/async (that would otherwise have to add an extra start call after upgrading)

Reply:
I see the issue, thanks for the example.
While **I fully agree for the module part**, I'm still not fond of having to change/add something to an existing setup to have it working after the upgrade, for the defer/async part.

I would disagree on "much less complex", since it seems more intuitive to me to check the docs for a parameter to disable htmx auto initialization (which is a behaviour I'd expect when importing any similar lib), rather than wondering why does the lib suddenly not work anymore when I set it as defer/async, especially given that [deferring your scripts avoids blocking rendering and improve page load times](https://developer.chrome.com/docs/lighthouse/performance/render-blocking-resources/#how-to-eliminate-render-blocking-scripts)

In that case, I'd say the most common usage would be to use htmx without any extension, deferring it because you follow Lighthouse (or any other tool)'s recommandations to improve page load times, yet with this change, htmx wouldn't work anymore, and you'd have to wonder why.

Indeed, you would have here to use a meta tag to set that parameter in the config, since, as you said, the config object cannot be used before htmx gets loaded, but [as the htmx doc says](https://htmx.org/api/#config) anyway, 
> Note that using a [meta tag](https://htmx.org/docs/#config) is the preferred mechanism for setting these properties.

Reply:
I understand your point, so I have rewritten the entire code to add a config setting `htmx.config.autoStart`. To disable autostart when used as a module, I changed the UMD wrapper and set the property `htmx.config.autoStart = false`. This way it is not necessary to check the `document.currentScript` property.

I have added another commit that I would like to discuss. The idea is to wait until all external extensions scripts are loaded before initializing htmx when `defer` or `async` is used. Therefore you need to mark all of the extension script elements with an attribute `hx-ext` (not really sure, if we should re-use this attribute). In the initialization phase of htmx, htmx now waits for the `onload` events of all extension script elements to fire and then initializes itself.

```html
<script src="htmx.js" defer></script>
<script src="ext/preload.js" defer hx-ext></script>
<script src="ext/debug.js" defer hx-ext></script>
```

Reply:
Thanks for the effort!

That scripts attribute system would work indeed, and wouldn't cause retro compatibility issues here because of the change to add to existing setups, since it wasn't working properly anyway

I was wondering though, when looking at [the htmx code](https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#LL3562C13-L3562C24), there is this setTimeout that lets any other listener bound to [DOMContentLoaded](https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L3498), do their logic before the `htmx:load` event gets fired

Couldn't we apply the same logic here before [processing the body](https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L3535) ?
This would make the following chain call

1. htmx executes its `ready` callback, [merging configs](https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L3532) with the meta tags, setups the [htmx:abort](https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L3539) listener, the [onpopstate](https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L3547) listener
2. call setTimeout with a delay of 0, so that the function returns here, and the rest of the code executes later
3. All other listeners bound to DOMContentLoaded in the document get called & processed. Here, extensions could bind themselves to htmx, after htmx has been initialized, but before htmx has processed the body
4. the setTimeout callback gets called right after, htmx processes the body

Do you think that would work? From the top of my head, I can't think of any reason it wouldn't, but I may be missing something here.
The advantage of that solution (given that it works), would be that you wouldn't even need to add the hx-ext attribute to extension scripts, nor wait for all of them to initialize before auto-starting htmx. If I get it correctly, it would work by itself, automatically. But again, I may be missing something since I haven't tried out what I'm suggesting here

Reply:
I like the idea and gave it a try. It only works for `defer`, but as you said, this is the preferred way to defer your scripts. In the few cases where you use async, you will need to adjust the configuration setting. I did a force-push to adjust the code.